File: 0050-powx-n/solution.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            return 1/(self.myPow(x, -1*n))
        elif n == 0:
            return 1
        elif n == 1:
            return x
        if n%2 == 0:
            X = self.myPow(x, n/2)
            return X*X
        else:
            X= self.myPow(x, n//2)
            return X*X*x
        
            
        

File: 0050-powx-n/solution.cpp
class Solution {
public:
    double myPow(double x, int n) {
        if (x == 1.0 || n == 0) {
            return 1.0;
        }

        if (n < 0 ) {
            if (n == INT_MIN) {
                return (1.0/myPow(x, INT_MAX)) * (1.0/x);
            }
            return 1.0/myPow(x, -n);
        } 

        double ans = 1.0f;
        double x_power = x;
        while (n) {
            if (n & 1) {
                ans *= x_power;
            }
            x_power *= x_power;
            n >>= 1;
        }

        return ans;
    }
};
